### PR TITLE
Adding ETag support for If-Match header

### PR DESCRIPTION
**Pull request by riv2q**
If-Match RFC: https://tools.ietf.org/html/rfc7232#section-3.1

sample of usage:
```
╰─$ curl -i -1 -H 'If-Match: "ec6632fd7da26584a7953727e320f28f92ac3b"' http://127.0.0.1:7000/api/service/v1/service-status
HTTP/1.0 412 PRECONDITION FAILED
ETag: "b794fc43a15de07a30d3a700f08c0805"
Date: Mon, 18 Dec 2017 10:03:15 GMT
Accept-Ranges: none
Connection: close
Server: Werkzeug/0.14.dev Python/3.5.2


╰─$ curl -i -1 -H 'If-Match: "b794fc43a15de07a30d3a700f08c0805"' http://127.0.0.1:7000/api/service/v1/service-status
HTTP/1.0 200 OK
Content-Type: application/json
Content-Length: 16
ETag: "b794fc43a15de07a30d3a700f08c0805"
Date: Mon, 18 Dec 2017 11:14:23 GMT
Accept-Ranges: none
Server: Werkzeug/0.14.dev Python/3.5.2

{"status": "OK"}
```

### CODE REVIEW

File: tests/test_wrappers.py [hunk 1]
```diff
@@ -551,6 +559,48 @@ def test_etag_response_mixin():
...
+    response.cache_control.must_revalidate = True
+    response.cache_control.max_age = 60
+    response.headers['Content-Length'] = len(response.get_data())
+    assert response.headers['Cache-Control'] in ('must-revalidate, max-age=60',
+                                                 'max-age=60, must-revalidate')
+
+    assert 'date' not in response.headers
+    env = create_environ()
+    env.update({
+        'REQUEST_METHOD':       'GET',
```
Comment (ThiefMaster) [inline tests/test_wrappers.py]: what's the reason for this rather weird spacing?

File: tox.ini [hunk 2]
```diff
@@ -27,7 +27,7 @@ deps =
     python-memcached
     watchdog
 
-    hypothesis: hypothesis
+    hypothesis: hypothesis=3.41.0
```
Comment (ThiefMaster) [inline tox.ini]: are you sure this shouldn't be `==`? (assuming it's passed to `pip install` by tox)

Comment (riv2q) [reply tox.ini]: Hey, I was trying to check why hypothesis fails on 3.6 last working build was in version ```3.41.0``` but this didn't help so I reverted it.

File: tests/test_wrappers.py (hunk 1 above)
Comment (riv2q) [reply tests/test_wrappers.py]: done

**Comment by mitsuhiko** [comment]
Closing in favor of #1223 which is basically the same thing sans reformatting.

**Pull request closed without merging**